src/core/coap/coap_message.hpp

```diff
@@ -379,18 +379,42 @@ class Message : public ot::Message
      */
     otError AppendUriQueryOption(const char *aUriQuery);
 
+    /**
+     * This method returns a pointer to the first option matching the given option number.
+     *
+     * @param[in]   aOption         Option number to look for.
+     */
+    const otCoapOption *GetFirstOptionMatching(otCoapOptionType aOption);
+
     /**
      * This method returns a pointer to the first option.
      *
      */
     const otCoapOption *GetFirstOption(void);
 
+    /**
+     * This method returns a pointer to the next option matching the given option number.
+     *
+     * @param[in]   aOption         Option number to look for.
+     */
+    const otCoapOption *GetNextOptionMatching(otCoapOptionType aOption);
```

Comment:
Maybe we should Remove `Get` from `GetFirstOption()`, `GetNextOption()`, so that it's clear these methods are not ordinary getters. Thoughts? @jwhui

Reply:
Another idea, if we're concerned about the state of the `otCoapMessage` object changing as options are iterated over… we have a routine that initialises a `otCoapOptionIterator` which contains the option iterator state and can scan back and forth without changing the state of the message object itself.  That iterator is then private to the caller manipulating it.

Reply:
I think it makes sense to introduce a coap option iterator.

Reply:
Question is, do we do that here, or should that be a separate PR to separate concerns?

Reply:
@jwhui @bukepo 

Had a crack at an iterator in https://github.com/vrtsystems/openthread/commits/feature/coap-msg-opt-iterator

I have not tested that code beyond seeing if it compiles as the hardware needed for testing is at work and the office is closing for the year (will be open again 6th January… I'm actually home a day earlier due to smoke-haze induced illness).

If that's getting close to the mark, we can make another pull request (and re-base this one). :-)

Reply:
I've cleaned up that code and made a separate PR: #4452.